enhancements/network/bgp-overview.md

```diff
@@ -0,0 +1,176 @@
+  - https://github.com/openshift/enhancements/pull/394
+---
+
+# OpenShift and BGP
+
+BGP (Border Gateway Protocol) is a dynamic routing protocol used to power a
+number of important networking use cases, including routing in the global
+internet.  For some helpful resources on learning more about BGP, see the
+[BGP Resources](#bgp-resources) section of this document.
+
+The purpose of this document is to serve as a high level overview of how BGP
+relates to OpenShift.  We will review a number of use cases, discuss how they
+could be supported, and link to existing related work.  Any future enhancements
+related to BGP should be able to rely on this one to provide some broader
+context and should continue to update this document as necessary.
+
+## Use Cases
+
+### L3 (Layer 3) Network Fabric
+
+The use case here does not require any changes to OpenShift.  It is still
+useful to call out as it is one use of BGP that is relevant to OpenShift,
+particularly with on premise clusters.
+
+It is possible to use BGP between the routers used to build the L3 network that
+OpenShift Nodes are connected to.
+
+### Dynamic Routing for Egress Traffic
+
+*First, an aside on terminology: This is referring to IP Routing, and
+not the OpenShift feature called Routing that works at the HTTP layer.  There
+are also OpenShift features that use the word "egress", but in this context, we
+only mean to refer to traffic that is leaving the cluster and not related to
+any other features that use that name.*
+
+Some networks may have more than one router accessible from Nodes for
+availability reasons.  In this case an administrator may want to use BGP down
+to the Nodes themselves so they are aware of the multiple routing options and
+potentially use ECMP (Equal Cost Multi Path) to distribute traffic
+across each of the routers.
+
+This capability is not provided by OpenShift today.  It may be possible as a
+custom configuration, but not with all network providers.  With OpenShift-SDN,
+it should be possible to run your own BGP routing daemon on the host via a
+`DaemonSet`.
+
+OVN-Kubernetes uses a different host network configuration where egress
+traffic is routed by Open vSwitch flows programmed by OVN.  It means that a
+dynamic routing daemon must also feed those routes into OVN configuration, so
+it's not as straight forward as just running an existing daemon on the host.
+
+A future enhancement is needed to describe how we would provide this
+functionality natively with OpenShift.
+
+### Pod Network Control Plane
+
+Many people become aware of BGP in the Kubernetes community via its use in
+[Calico][1].  The primary way BGP is used in Calico is as a control plane for
+the Pod network.  In this case, the use of BGP should be considered an
+implementation detail of the control plane.
+
+You may read Calico's own documentation to understand more details about how
+Calico uses BGP.
+
+There are other ways this control plane functionality can be implemented, such
```

Comment:
Can you please clarify what do you mean by "control plane for pods networks" in this section? I don't have much experience with using Calico, but at least for on-premises deployments, it looks like Calico lets you peer directly with the physical network fabric using iBGP or eBGP and advertise routes. I don't think this could be done by the Kubernetes API, but it's true that they could have chosen a different dynamic routing protocol for this: https://www.projectcalico.org/why-bgp/

Reply:
Calico uses BGP for many things, depending on configuration options - all of which are covered later in this document.

@russellb I suspect you're going to have to call this out more specifically, given how confusing this space is. "Note that Calico uses BGP **both** as a control plane and to implement some of the features below, **but these are actually independent**. A BGP control plane is **not**  a prerequisite."

Reply:
Yeah, this is confusing, and maybe mentioned Calico specifically is just making the confusion worse.

The use case I'm trying to call out here is that *somehow*, the system must know the physical location for each Pod network IP address.  BGP is one mechanism which can be used to accomplish this.

Maybe it would be helpful to describe some of the alternatives: OpenShift-SDN uses information in the Kubernetes API and OVN has a "Port_Bindings" table in its southbound database.

Reply:
I've updated this section.  Take a look and see if it's better.  I dropped mention of Calico, because I think it was making things more confusing.

Reply:
Thanks @russellb. LGTM now

Reply:
I like @squeed's suggestion above. If we say the goal of this doc is to make it possible for a wider group of people to engage meaningfully in an "OpenShift needs BGP support" discussion, I think there needs to be some breakdown of how BGP is used to implement these features in other solutions (in particular Calico, which for many has become synonymous with "BGP support").

Reply:
Thanks for the feedback!  I've re-introduced some Calico references in a couple of sections and included a new Calico focused section at the end.  I haven't gone through and mapped Calico docs to every use case, so the Calico information is probably incomplete.  I'm sure we could map it to more use cases, but this is a start at least.